dm/pkg/checker/table_structure.go

```diff
@@ -267,114 +311,207 @@ func (c *TablesChecker) checkTableOption(opt *ast.TableOption) *incompatibilityO
+			}
 
-				ctStmt, ok := stmt.(*ast.CreateTableStmt)
-				if !ok {
-					markCheckError(r, errors.Errorf("Expect CreateTableStmt but got %T", stmt))
-					r.Extra = fmt.Sprintf("instance %s on sharding %s", instance, c.name)
-					return r
+			info, err := dbutil.GetTableInfoBySQL(statement, parser2)
+			if err != nil {
+				if len(errCh) == 0 {
+					errCh <- err
 				}
-
-				if c.checkAutoIncrementPrimaryKey {
-					passed := c.checkAutoIncrementKey(instance, schema, table, ctStmt, info, r)
-					if !passed {
-						return r
-					}
+				return
+			}
+			stmt, err := parser2.ParseOneStmt(statement, "", "")
+			if err != nil {
+				if len(errCh) == 0 {
+					errCh <- errors.Annotatef(err, "statement %s", statement)
 				}
+				return
+			}
 
-				if stmtNode == nil {
-					stmtNode = ctStmt
-					firstTable = dbutil.TableName(schema, table)
-					firstInstance = instance
-					continue
+			ctStmt, ok := stmt.(*ast.CreateTableStmt)
+			if !ok {
+				if len(errCh) == 0 {
+					errCh <- errors.Errorf("Expect CreateTableStmt but got %T", stmt)
 				}
+				return
+			}
 
-				checkErr := c.checkConsistency(stmtNode, ctStmt, firstTable, dbutil.TableName(schema, table), firstInstance, instance)
-				if checkErr != nil {
-					r.State = StateFailure
-					r.Errors = append(r.Errors, checkErr)
-					r.Extra = fmt.Sprintf("error on sharding %s", c.name)
-					r.Instruction = "please set same table structure for sharding tables"
-					return r
+			if c.checkAutoIncrementPrimaryKey {
+				passed := c.checkAutoIncrementKey(instance, table.Schema, table.Name, ctStmt, info, r)
+				if !passed {
+					return
 				}
 			}
+
+			checkErr := c.checkConsistency(stmtNode, ctStmt, firstTable, table.String(), firstInstance, instance)
+			if checkErr != nil {
+				r.State = StateFailure
+				r.Errors = append(r.Errors, checkErr)
+				r.Extra = fmt.Sprintf("error on sharding %s", c.targetTableID)
+				r.Instruction = "please set same table structure for sharding tables"
+				return
+			}
+		}
+	}
+
+	for instance, tables := range c.tableMap {
+		concurrency := getConcurrency(len(tables))
+		maxConnections, err := utils.GetMaxConnections(ctx, c.dbs[instance])
+		if err != nil {
+			return r, err
 		}
+		if concurrency > maxConnections {
+			concurrency = maxConnections
+		}
+		onethread := len(tables) / concurrency
+		log.L().Logger.Info("this source have", zap.String("source", instance), zap.Int("concurrency", concurrency), zap.Int("onthread check num", onethread), zap.Int("table num", len(tables)))
+		for i := 0; i < concurrency; i++ {
+			if onethread == 0 {
+				checkWg.Add(1)
+				go checkFunc(instance, tables[0:])
+				break
+			}
+			if onethread*i >= len(tables) {
+				break
+			}
+			var checkTables []*filter.Table
+			if i+1 == concurrency || onethread*(i+1) >= len(tables) {
+				checkTables = tables[onethread*i:]
+			} else {
+				checkTables = tables[onethread*i : onethread*(i+1)]
+			}
+			checkWg.Add(1)
+			go checkFunc(instance, checkTables)
+		}
+	}
```

Comment:
Why not just determine the value of concurrency and then use something like the producer-consumer model?

Reply:
> use something like the producer-consumer model?

I haven't written many concurrent programs before... So I just wrote it according to my own feelings.

I think there's not much difference between them. Now it is equivalent to passing all the tables that need to be consumed to consumers at one time.

If you think producer-consumer model is better, I can update this to producer-consumer model..

Reply:
agrees with GMHDBJD `producer-consumer model`  is good for maintenance eg:

if i want to change the concurent count , i just incr the worker count
if i want to change the dispatch job logic, i just need to change the producer side
if i want to change the job handle logic, i just need to c change the worker side logic